VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+        if (curTime.year != info.issueYear || curTime.month != (info.issueMonth + 1)) {
+            revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+        } else if (curTime.month == 1) {
+            if (curTime.year != (info.issueYear + 1) || info.issueMonth != 12) {
+                revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+            }
+        }
+    }
+
+    /**
+     * @notice Validate proof signer
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address, Used to revert
+     * @param proofSigner - Signer of proof
+     * @param info - Claim information
+     */
+    function _validateProofSigner(address didAddress, address proofSigner, ClaimInfo calldata info) internal virtual view {
+        bool isVerified;
+        uint index;
+        uint length = _trustedSigners.length();
+
+        while (index < length && !isVerified) {
+            if (proofSigner == _trustedSigners.at(index)) {
+                isVerified = true;
+            }
+            unchecked { ++index; }
+        }
+
+        if (!isVerified) {
+            revert InvalidProof(didAddress, info.typeId, info.xp, info.signature);
+        }
+    }
+
+    /**
+     * @notice Check the proofs and return the total amount of XP
+     * @param didAddress - DID address
+     * @param infos - Array of claim requests
+     * @return uint - Total amount of requested claim
+     */
+    function _validateClaimSignature(address didAddress, ClaimInfo[] calldata infos) internal virtual returns(uint) {
+        uint totalXP;
+
+        bytes memory rawMsg;
+        bytes32 dataHash;
+        address proofSigner;
+
+        uint infoLen = infos.length;
+        DateTime._DateTime memory curTime = DateTime.parseTimestamp(block.timestamp);
+
+        for (uint i; i < infoLen;) {
+            if (infos[i].xp == 0) {
+                revert InvalidXP(didAddress, infos[i].typeId, infos[i].xp);
+            }
+            _validateProofTime(didAddress, curTime, infos[i]);
+
+            rawMsg = abi.encodePacked(didAddress, infos[i].typeId, infos[i].issueYear, infos[i].issueMonth);
+            
+            // Check whether `typeId` is claimed in this month
+            if (isClaimedTypeId[rawMsg]) {
+                revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+            }
+            isClaimedTypeId[rawMsg] = true;
+
+            if (bytes(infos[i].uniqueId).length != 0) {
+                rawMsg = abi.encodePacked(infos[i].typeId, infos[i].uniqueId);
+                // Check whether `uniqueId` is claimed before
+                if (isClaimedUniqueId[rawMsg]) {
+                    revert DuplicatedUniqueId(didAddress, infos[i].typeId, infos[i].uniqueId, infos[i].xp);
+                }
+                isClaimedUniqueId[rawMsg] = true;
```

Comment:
Should this only be set at the very end if all the tests pass?

Reply:
Same as above.
Checks the `uniqueId` in each for loop.

Reply:
As above

Reply:
Same as above.